csrc/multidevice/utils.cpp

```diff
@@ -266,40 +235,65 @@ void insertReshardings(Fusion* fusion) {
         expr->outputs().size() == 1,
         "multi-output expressions are not supported");
     auto output = expr->outputs().at(0)->as<TensorView>();
-    std::vector<TensorView*> new_inputs;
-    for (auto input : getTvsWithDifferentSharding(
-             output, ir_utils::filterByType<TensorView>(expr->inputs()))) {
-      // TODO: reuse cacheAfter?
-      // TODO: here we should add a mechanism to potentially reuse the inserted
-      // resharding accross all the consumer of the resharded tensor. This way
-      // we could avoid wasteful resharding set insertion.
-      TensorView* new_input = set(input);
-      new_inputs.push_back(new_input);
-      expr = ir_utils::replaceValInExprInputs(expr, input, new_input);
+
+    auto inputs = getTvsWithDifferentSharding(
+        output, ir_utils::filterByType<TensorView>(expr->inputs()));
+
+    // Insert resharding set after the expr when there is only one input.
+    // input [expr] output [set] new_output
+    if (!inputs.empty() && expr->inputs().size() == 1) {
+      auto input = *inputs.begin();
+      TensorView* new_output = set(output);
+      auto new_output_map = ir_utils::replaceValInAllExprInputsAndFusionOutputs(
```

Comment:
`new_output->definition()` is also a use of `output` at the point. Does replaceValInAllExprInputsAndFusionOutputs also replace `output` in **that** set, leading to a self-loop?

Reply:
Nope! I printed out what expressions get modified and it did not include the set. 

I think a TensorView's `uses()` doesn't get recalculated while updating use info? https://github.com/NVIDIA/Fuser/blob/72bc243dfa5e4a7130d80e04d76fed8b4cf5f842/csrc/ir/base_nodes.cpp#L92

TBH, I don't know if that check was intended to cover my use case, or I just got lucky and it does?

Reply:
That's bizarre :) Let me run a quick test and then update this thread.

Reply:
During the reset, https://github.com/NVIDIA/Fuser/blob/dad05c3f5fc6eee85057380cdd0643ee84bb0475/csrc/fusion.cpp#L617 only looks at "reachable" expressions so the use of `output` in `new_output->definition()` (the `set`) is not replaced. So it's safe to use `replaceValInAllExprInputsAndFusionOutputs` here. TIL!